rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,112 @@
+---
+feature: dynamic-ids
+start-date: 2019-09-05
+author: Silvan Mosberger
+co-authors: (find a buddy later to help our with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/65698
+---
+
+# Summary
+[summary]: #summary
+
+A lot of NixOS modules are [assigning static uids/gids](https://github.com/NixOS/nixpkgs/blob/044cc701c23ede96355eb1dc997985b4dfac0372/nixos/modules/misc/ids.nix#L36) to their users. This has resulted in less than 90 static ids left in the reserved range from 0 to 400.
+
+This RFC deprecates the practice of doing that and instead suggests to
+- If applicable use systemd's `DynamicUser`
```

Comment:
To help the reader to decide when not to use `DynamicUser`:
Services not applicable for `DynamicUser` are services that own many files,
so that the service startup would be significantly delayed by adapting filesystem permission.

Reply:
From http://0pointer.net/blog/dynamic-users-with-systemd.html
> the allocation of dynamic UIDs has been tweaked in two ways to avoid the necessity to do this expensive operation in most cases: firstly, when a dynamic UID is allocated for a service an allocation loop is employed that starts out with a UID hashed from the service's name. This means a service by the same name is likely to always use the same numeric UID. That means that a stable service name translates into a stable dynamic UID, and that means recursive file ownership adjustments can be skipped (of course, after validation). Secondly, if the configured state directory already exists, and is owned by a suitable currently unused dynamic UID, it's preferably used above everything else, thus maximizing the chance we can avoid the chown()ing.

So I'm not sure how much of an issue that is in practice without benchmark, but it sounds like it would work even for this case.

Reply:
I think this should be good enough for us.